File: 2.AddTwoNumbers/c/1s_addtwonumbers.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

struct ListNode 
{
    int val;
    struct ListNode* next;
};


/**
 * @brief Adds two ListNodes.
 *
 * @param l1 
 * @param l2
 *
 * @return Resulting ListNode.
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2);

/**
 * @brief Instantiates a new ListNode.
 *
 * @param val value of ListNode.
 *
 * @return Instantiated ListNode.
 */
struct ListNode* ListNode_new(int val);

/**
 * @brief Adds a ListNode to a ListNode.
 *
 * @param l     set its next
 * @param val   value of next
 *
 * @return Pointer to instantiated ListNode.
 */
struct ListNode* ListNode_add(struct ListNode* l, int val);

/**
 * @brief Deallocates memory used by a ListNode.
 *
 * @param l 
 */
void ListNode_del(struct ListNode* l);

/**
 * @brief Prints a ListNode.
 *
 * @param l
 */
void ListNode_print(struct ListNode* l);


int main()
{
    struct ListNode* l1 = ListNode_new(2);
    struct ListNode* l2 = ListNode_add(l1, 4);
    struct ListNode* l3 = ListNode_add(l2, 3);

    struct ListNode* r1 = ListNode_new(5);
    struct ListNode* r2 = ListNode_add(r1, 6);
    struct ListNode* r3 = ListNode_add(r2, 4);

    struct ListNode* res = addTwoNumbers(l1, r1);
    printf("  ");
    ListNode_print(l1); 
    printf("+ ");
    ListNode_print(r1); 
    printf("= ");
    ListNode_print(res);

    ListNode_del(l1);
    ListNode_del(r1);
    ListNode_del(res);

    return 0;
}


struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2)
{
    struct ListNode* result = ListNode_new(0);
    struct ListNode* node = result;
    int carry = 0;
    for(;;)
    {
        int l1_n = 0;
        int l2_n = 0;
        int x = 0;
        int y = 0;

        if (l1 != NULL)
        {
            x = l1->val;
            l1 = l1->next;
            l1_n = 1;
        }
        if (l2 != NULL)
        {
            y = l2->val;
            l2 = l2->next;
            l2_n = 1;
        }

        int sum = x + y + carry;
        if (sum == 0 && l1_n == 0 && l2_n == 0)
        {
            struct ListNode* res = result->next;
            free(result);
            return res;
        }

        if (sum > 9)
        {
            sum %= 10;
            carry = 1;
        }
        else
        {
            carry = 0;
        }

        ListNode_add(node, sum);
        node = node->next;
    }
}


struct ListNode* ListNode_new(int val)
{
    struct ListNode* l = malloc(sizeof(struct ListNode));
    l->val = val;
    l->next = NULL;
    return l;
}


struct ListNode* ListNode_add(struct ListNode* l, int val)
{
    l->next = ListNode_new(val);
    return l->next;
}


void ListNode_del(struct ListNode* l)
{
    if (l->next != NULL)
        ListNode_del(l->next);
    free(l);
}


void ListNode_print(struct ListNode* l)
{
    struct ListNode* ite = l;
    while (ite != NULL)
    {
        printf("%d -> ", ite->val);
        ite = ite->next;
    }   
    printf("NULL\n");
}

File: 2.AddTwoNumbers/python/1s_addtwonumbers.py
from typing import List


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def addNode(l: ListNode, v: int) -> ListNode:
    """
    Sets the next node for l.
    """
    node = ListNode(v)
    l.next = node
    return l


def printListNode(l: ListNode) -> None:
    node = l
    while node is not None:
        print(node.val, end=" -> ")
        node = node.next
    print("None")


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        """
        Adds two ListNodes and returns the result as a ListNode.
        """
        result = ListNode(0)
        node = result
        carry = 0
        while True:
            if not l1 is None:
                x = l1.val
                l1 = l1.next
                l1_n = True
            else:
                x = 0
                l1_n = False
            if not l2 is None:
                y = l2.val
                l2 = l2.next
                l2_n = True
            else:
                y = 0
                l2_n = False
            res = x + y + carry
            if res == 0 and not l1_n and not l2_n:
                return result.next
            if res > 9:
                res = res % 10
                carry = 1
            else:
                carry = 0

            addNode(node, res)
            node = node.next


def genExamples() -> List[ListNode]:
    l1 = ListNode(3)
    l2 = ListNode(4)
    l3 = ListNode(2)
    l2.next = l1
    l3.next = l2
    r1 = ListNode(4)
    r2 = ListNode(6)
    r3 = ListNode(5)
    r2.next = r1
    r3.next = r2
    return [r3, l3]


def main():
    solution = Solution()
    examples = genExamples()
    result = solution.addTwoNumbers(examples[0], examples[1])
    print('  ', end='')
    printListNode(examples[0])
    print('+ ', end='')
    printListNode(examples[1])
    print('= ', end='')
    printListNode(result)


if __name__ == "__main__":
    main()
